server/datastore/mysql/policies.go

```diff
@@ -96,6 +96,60 @@ func (ds *Datastore) SavePolicy(ctx context.Context, p *fleet.Policy) error {
 	return nil
 }
 
+// NewFailingPolicies returns a list of new failing policies. By "new" we mean those that fail on their first
+// run and those that were passing on the previous run and are failing on the incoming execution.
+func (ds *Datastore) NewFailingPoliciesForHost(
+	ctx context.Context,
+	hostID uint,
+	incomingResults map[uint]*bool,
+) ([]uint, error) {
+	// Sort the results to have generated SQL queries ordered to minimize
+	// deadlocks. See https://github.com/fleetdm/fleet/issues/1146.
+	orderedIDs := make([]uint, 0, len(incomingResults))
+	for policyID := range incomingResults {
+		orderedIDs = append(orderedIDs, policyID)
+	}
+	sort.Slice(orderedIDs, func(i, j int) bool {
+		return orderedIDs[i] < orderedIDs[j]
+	})
+
+	selectQuery := fmt.Sprintf(
+		`SELECT policy_id, passes FROM policy_membership
+		WHERE host_id = ? AND policy_id IN (?)`,
+	)
+	var fetchedPolicyResults []struct {
+		ID     uint  `db:"policy_id"`
+		Passes *bool `db:"passes"`
+	}
+	selectQuery, args, err := sqlx.In(selectQuery, hostID, orderedIDs)
+	if err != nil {
+		return nil, ctxerr.Wrapf(ctx, err, "build select policy_membership query")
+	}
+	if err := sqlx.SelectContext(ctx, ds.reader, &fetchedPolicyResults, selectQuery, args...); err != nil {
+		return nil, ctxerr.Wrapf(ctx, err, "select policy_membership")
+	}
+	prevPolicyResults := make(map[uint]bool)
+	for _, result := range fetchedPolicyResults {
+		prevPolicyResults[result.ID] = *result.Passes
+	}
+	// Loop through latest results, collecting which policies are going from
+	// pass to fail and those that fail in the first execution.
+	var newFailingPolicies []uint
+	for incomingPolicyID, incomingResult := range incomingResults {
+		if incomingResult == nil || *incomingResult {
```

Comment:
Just a question as much for me as for the review, I thought that nil result meant the same as false/failure? This treats it as success.

Reply:
Good catch, from the ingestion it looks like: 
1. incomingResult == nil: failed to run policy query.
2. incomingResult != nil && *incomingResult: policy run and succeeded.
3. incomingResult != nil && !*incomingResult: policy run and failed (no rows).

Will fix.

Reply:
:+1: thanks for the clarification, wasn't sure myself, the semantics you describe make a lot of sense.

Reply:
Updated the code after Noah's feedback. For webhooks we only want to consider a policy as "failed" when it ran successfully and it didn't return any rows.